XXX-asyncwrap-api.md

```diff
@@ -0,0 +1,309 @@
+| Title  | AsyncWrap API |
+|--------|---------------|
+| Author | @trevnorris   |
+| Status | DRAFT         |
+| Date   | 2016-04-05    |
+
+## Description
+
+Since its initial introduction with the `AsyncListener` API, `AsyncWrap` has
+slowly evolved to ensure a generalized API that would serve as a solid base for
+module authors who wished to add events to the life cycle of the event loop.
+The API presented below aims to accomplish this.
+
+## API
+
+### Overview
+
+```js
+// Standard way of requiring a module. Snake case follows core module
+// convention.
+const async_wrap = require('async_wrap');
+
+// List of asynchronous providers (e.g. CRYPTO or TCPWRAP). Can be used for
+// trace filtering of events.
+const providers = async_wrap.Providers;
+
+// init() is called during object construction. The object handle being
+// constructed can be accessed as "this". Because init() is called during
+// construction it is not safe to blindly inspect the constructed object at the
+// moment. Though we are currently investigating how to make this safe.
+function init(id, type, parentId) { }
+
+// pre() is called just before the handle's callback is called. It can be
+// called 0-N times for handles (e.g. TCPWRAP), and should be called exactly 1
+// time for requests (e.g. FSREQWRAP).
+function pre(id) { }
+
+// post() is called just after the handle's callback has finished, and will be
+// called regardless whether the handle's callback threw. If the handle's
+// callback did throw then hasThrown will be true.
+function post(id, hasThrown) { }
+
+// destroy() is called when an AsyncWrap instance is destroyed. In cases like
+// HTTPPARSER where the resource is reused, or timers where the handle is only
+// a JS object, destroy() will be triggered manually soon after post() has
+// completed. This is the only callback that does not receive the context as
+// "this". Because it is called while the associated class is being destructed
+// it is unsafe to operate on the object.
+function destroy(id) { }
+
+// Add a set of hooks to be called during the lifetime of asynchronous events.
+// createHook() returns an instance of the AsyncHook constructor that can
+// control how those hooks are used. For example, enabling or disabling their
+// execution.
+asyncHook = async_wrap.createHook({ init, pre, post, destroy });
+
+// Allow callbacks of this AsyncHook instance to fire. This is not an implicit
+// action after running createHook(), and must be explicitly run to being
+// executing callbacks.
+asyncHook.enable();
+
+// Disable listening for new asynchronous events. Though this will not prevent
+// callbacks from firing on asynchronous chains that have already run within
+// the scope of an enabled AsyncHook instance.
+asyncHook.disable();
+
+// Enable hooks for the current synchronous execution scope. This will ensure
+// the hooks are not in effect in case of multiple returns, or if an exception
+// is thrown.
+asyncHook.scope();
```

Comment:
Running the following:
```js
hook.enable();
net.createServer(fn).listen(8080);
hook.disable();
```
will allow `hook`'s callback to continue firing on all callbacks that are triggered by the new `Server` instance. Though this can be error prone if there are multiple return locations. e.g.
```js
setTimeout(() => {
  hooks.enable();
  if (checkState())
    return;
  fs.open(path, r, fn);
  hooks.disable();
}, 100);
```
Because of the early return and not first running `hooks.disable()` that set of hooks could remain enabled indefinitely. So `hooks.scope()` would guarantee that they are disabled when the current stack unwinded. e.g.
```js
setTimeout(() => {
  hooks.scope();
  if (checkState())
    return;
  fs.open(path, r, fn);
}, 100);
```
I'll probably end up needing to place in massive bold letters that it lasts until the **current stack has unwound**, or else users will think it's function scope specific. Like `const` or `let`.